### PR TITLE
Update for pytorch v0.4.1

### PR DESCRIPTION
**Pull request by kan-bayashi**
#312

### CODE REVIEW

**Comment by kan-bayashi** [comment]
Now testing...
- [ ] ljspeech / single gpu
- [ ] wsj / single gpu
- [ ] ljspeech multi gpu
- [ ] wsj /multi gpu

File: test/test_loss.py [hunk 1]
```diff
@@ -32,12 +32,11 @@ def test_ctc_loss():
...
                         for x in np_pred], 0.0).transpose(0, 1)
-    th_target = torch.autograd.Variable(
-        torch.from_numpy(numpy.concatenate(np_target)))
-    th_ilen = torch.autograd.Variable(torch.from_numpy(input_length))
-    th_olen = torch.autograd.Variable(torch.from_numpy(label_length))
+    th_target = torch.from_numpy(numpy.concatenate(np_target))
+    th_ilen = torch.from_numpy(input_length)
+    th_olen = torch.from_numpy(label_length)
     # NOTE: warpctc_pytorch.CTCLoss does not normalize itself by batch-size while chainer's default setting does
     th_loss = (CTCLoss()(th_pred, th_target, th_ilen,
```
Comment (ShigekiKarita) [inline test/test_loss.py]: use CTCLoss(size_averaged=True) instead

Comment (ShigekiKarita) [reply test/test_loss.py]: maybe adding gradient equality check is nice. the different gradient was the reason why we used custom function? https://github.com/jnishi/warp-ctc/commit/3fd4783497bc1791034679ba70d1c74efc4a4468

Comment (ShigekiKarita) [reply test/test_loss.py]: also check this https://github.com/espnet/espnet/pull/307

Comment (ShigekiKarita) [reply test/test_loss.py]: For example wrapping with 
``` python
class SquareGrad(Function):
    @staticmethod
    def forward(ctx, input):
        return x
    @staticmethod
    def backward(ctx, grad_output):
        return grad_output * grad_output

def chainer_like_ctc(y_hat, y_true, ilens, olens):
    return warp_ctc.CTCLoss(size_average=True)(SquareGrad.apply(y_hat), y_true, ilens, olens)
```

Comment (kan-bayashi) [reply test/test_loss.py]: Thanks @ShigekiKarita.
Here the return value `x` in forward is `input`, right?

**Comment by ShigekiKarita** [comment]
LGTM

**Comment by kan-bayashi** [comment]
@ShigekiKarita if we use `chainer_like_ctc_loss`, `test_loss_and_ctc_grad` is failed.
If we use `CTCLoss(size_average=True)`, the test is OK.
This means that we don't have to modify CTC. 
Is my understanding is correct?

File: tools/Makefile [hunk 2]
```diff
@@ -31,10 +31,10 @@ nkf:
 	cd nkf; tar zxvf nkf-2.1.4.tar.gz; cd nkf-2.1.4; $(MAKE) prefix=.
 
 venv/lib/python2.7/site-packages/torch: venv/bin/activate
-	. venv/bin/activate; pip install pip --upgrade; pip install torch==0.4.0
+	. venv/bin/activate; pip install pip --upgrade; pip install torch==0.4.1
 
 warp-ctc: venv/lib/python2.7/site-packages/torch
-	git clone https://github.com/jnishi/warp-ctc.git
+	git clone https://github.com/SeanNaren/warp-ctc.git
```
Comment (jnishi) [inline tools/Makefile]: Use  `follows-upstream` branch of https://github.com/jnishi/warp-ctc of supports  because it supports pytorch 0.4.1 and https://github.com/espnet/espnet/pull/307.

File: .travis.yml [hunk 3]
```diff
@@ -50,8 +42,8 @@ install:
   # install matplotlib
   - pip install matplotlib
   # install warp-ctc
-  - git clone https://github.com/SeanNaren/warp-ctc.git 
-  - cd warp-ctc && git checkout 9e5b238f8d9337b0c39b3fd01bbaff98ba523aa5 && mkdir build && cd build && cmake .. && make -j4 && cd ..
+  - git clone https://github.com/SeanNaren/warp-ctc.git
```
Comment (jnishi) [inline .travis.yml]: As I noted in `Makefile`, use follows-upstream branch of https://github.com/jnishi/warp-ctc of supports because it supports pytorch 0.4.1 and #307.

File: tools/Makefile (hunk 2 above)
Comment (jnishi) [reply tools/Makefile]: When this PR is merged, I will fix `master` branch to support pytorch 0.4.1.

**Comment by kan-bayashi** [comment]
I will re-create pull request for dev branch.
